pkl-core/src/main/java/org/pkl/core/externalreader/ExternalReaderProcessImpl.java

```diff
@@ -49,14 +49,15 @@ public class ExternalReaderProcessImpl implements ExternalReaderProcess {
   private final Map<String, Future<@Nullable ResourceReaderSpec>>
       initializeResourceReaderResponses = new ConcurrentHashMap<>();
 
-  private @GuardedBy("this") boolean closed = false;
+  private final Object lock = new Object();
+  private @GuardedBy("lock") boolean closed = false;
```

Comment:
Encapsulating the locking protocol. Otherwise, users of the class (via the interface) might cause deadlocks by themselves synchronizing on `this`. You can find a better explanation in Effective Java 3rd Ed, Item 82.

Reply:
Gotcha, thanks!